tsl/src/continuous_aggs/materialize.c

```diff
@@ -950,48 +1069,77 @@ update_materializations(SchemaAndName partial_view, SchemaAndName materializatio
 }
 
 static bool
-get_invalidation_range(List *invalidations, InternalTimeRange *invalidation_range)
+ranges_overlap(InternalTimeRange invalidation_range, InternalTimeRange new_materialization_range)
 {
-	bool has_invalidations = false;
-	ListCell *lc;
+	Assert(invalidation_range.start <= invalidation_range.end);
+	Assert(new_materialization_range.start <= new_materialization_range.end);
+	return !(invalidation_range.end < new_materialization_range.start ||
+			 new_materialization_range.end < invalidation_range.start);
+}
 
-	invalidation_range->start = PG_INT64_MAX;
-	invalidation_range->end = PG_INT64_MIN;
-	foreach (lc, invalidations)
-	{
-		Invalidation *i = (Invalidation *) lfirst(lc);
-		Assert(i->lowest_modified_value <= i->greatest_modified_value);
-		if (i->lowest_modified_value < invalidation_range->start)
-			invalidation_range->start = i->lowest_modified_value;
-		if (i->greatest_modified_value > invalidation_range->end)
-			invalidation_range->end = i->greatest_modified_value;
-
-		has_invalidations = true;
-	}
+/* Checks to ensure a valid length */
+static void
+range_check(const InternalTimeRange range, int64 bucket_width)
+{
+	if (range.start > range.end)
+		elog(ERROR, "internal error: range start > end");
+
+	/* empty range does not need to be time-bucketed */
+	if (range_length(range) == 0)
+		return;
+
+	if (range.start != PG_INT64_MIN &&
+		range.start != ts_time_bucket_by_type(bucket_width, range.start, range.type))
+		elog(ERROR, "internal error: range start not aligned on bucket boundary");
 
-	return has_invalidations;
+	if (range.end != ts_time_bucket_by_type(bucket_width, range.end, range.type))
+		elog(ERROR, "internal error: range end not aligned on bucket boundary");
 }
 
-static void
-time_bucket_range(InternalTimeRange *range, int64 bucket_width)
+static int64
+range_length(const InternalTimeRange range)
 {
-	range->start = ts_time_bucket_by_type(bucket_width, range->start, range->type);
-	range->end = ts_time_bucket_by_type(bucket_width, range->end, range->type);
+	Assert(range.end >= range.start);
+
+	return int64_saturating_sub(range.end, range.start);
+}
 
-	if (range->end < PG_INT64_MAX - bucket_width)
-		range->end += bucket_width;
+/* Expand the range `range` to the right so that it covers as much of
+ * `to_cover` as possible while still obeying `max_length` */
+static bool
+range_expand_right(InternalTimeRange *range, const InternalTimeRange to_cover, int64 max_length)
+{
+	int64 max_end = int64_saturating_add(range->start, max_length);
 
-	if (range->start > range->end)
-		range->start = range->end;
+	Assert(range->start <= to_cover.start);
+	if (max_end > to_cover.start)
+	{
+		range->end = int64_min(to_cover.end, max_end);
```

Comment:
This does not seem to time_bucket the end correctly?

Reply:
good catch fixed